Introduce a common base response class to all single doc write ops

bleskes opened the pull request:
IndexResponse, DeleteResponse and UpdateResponse share some logic. This can be unified to a single DocWriteResponse base class. On top, some replication actions are now not about write operations anymore. This commit renames ActionWriteResponse to ReplicationResponse

Last some toXContent is moved from the Rest layer to the actual response classes, for more code re-sharing.

This is ported over from the `feature/seq_no` branch to make maintaining that branch simpler

bleskes commented:
@ywelsch can you please take a look?

ywelsch commented on core/src/main/java/org/elasticsearch/action/DocWriteResponse.java:
```diff
@@ -0,0 +1,122 @@
+/**
+ * A base class for the response of a write operation that involves a single doc
+ */
+public abstract class DocWriteResponse extends ReplicationResponse implements ToXContent {
```
call this DocOperationResponse (as it includes deletes)?

bleskes replied:
I think deleted is a write operation? DocOperationResponse as a name does cover Get, which isn't part of this game...

ywelsch commented on core/src/test/java/org/elasticsearch/update/UpdateIT.java:
```diff
@@ -19,6 +19,7 @@
 
 package org.elasticsearch.update;
 
+import com.carrotsearch.randomizedtesting.annotations.Repeat;
```
remove...

ywelsch commented on core/src/main/java/org/elasticsearch/rest/action/delete/RestDeleteAction.java:
```diff
@@ -65,15 +65,10 @@ public void handleRequest(final RestRequest request, final RestChannel channel,
             @Override
             public RestResponse buildResponse(DeleteResponse result, XContentBuilder builder) throws Exception {
-                ActionWriteResponse.ShardInfo shardInfo = result.getShardInfo();
-                builder.startObject().field(Fields.FOUND, result.isFound())
```
The method toXContent that writes out "found" is missing in DeleteResponse

bleskes replied:
good catch!

ywelsch commented on core/src/main/java/org/elasticsearch/action/index/IndexResponse.java:
```diff
@@ -110,12 +70,23 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
```
why is there only a toString() here, but not in DeleteResponse and UpdateResponse?

bleskes replied:
that's how it used to be.. I'll add.

ywelsch commented on core/src/main/java/org/elasticsearch/rest/action/bulk/RestBulkAction.java:
```diff
@@ -97,49 +98,35 @@ public RestResponse buildResponse(BulkResponse response, XContentBuilder builder
+                        builder.field(Fields._INDEX, itemResponse.getIndex());
+                        builder.field(Fields._TYPE, itemResponse.getType());
+                        builder.field(Fields._ID, itemResponse.getId());
                         builder.field(Fields.STATUS, itemResponse.getFailure().getStatus().getStatus());
```
I guess it was a deliberate choice not to return version = -1 in case of error.

bleskes replied:
we know it's -1 here:

```
    public long getVersion() {
        if (failure != null) {
            return -1;
        }

```

ywelsch commented on core/src/main/java/org/elasticsearch/rest/action/bulk/RestBulkAction.java:
```diff
@@ -97,49 +98,35 @@ public RestResponse buildResponse(BulkResponse response, XContentBuilder builder
+                                status = RestStatus.CREATED;
                             }
                         }
+                        builder.field(Fields.STATUS, status.getStatus());
```
can we add an abstract method getRestStatus() to DocWriteResponse (to be overwritten by its subclasses) so that we don't need all these instanceof...

bleskes replied:
good idea

martijnvg commented on core/src/main/java/org/elasticsearch/rest/action/index/RestIndexAction.java:
```diff
@@ -103,15 +103,9 @@ public void handleRequest(final RestRequest request, final RestChannel channel,
             @Override
             public RestResponse buildResponse(IndexResponse response, XContentBuilder builder) throws Exception {
```
Maybe we can let `DocWriteResponse` implement `StatusToXContent` instead of `ToXContent`? Then we can simply use `RestToXContentListener` as listener instead of an ananoumus inner class.

bleskes replied:
Good idea.

ywelsch commented on core/src/main/java/org/elasticsearch/action/DocWriteResponse.java:
```diff
@@ -0,0 +1,122 @@
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        ReplicationResponse.ShardInfo shardInfo = getShardInfo();
+        builder.field(Fields._INDEX, getIndex())
+            .field(Fields._TYPE, getType())
```
More of a personal preference: I don't like getters being used in place where fields can directly be accessed, especially if the fields are defined in the same class. Especially in case of heavy subclassing (and possibly dynamic dispatch on getters) this makes the superclass rely on behavior of subclass.

ywelsch commented:
Minor comments, o.w. LGTM

bleskes commented:
@ywelsch @martijnvg I addressed all your feedback and rebased on master (there was a big change to replication action there). Can you take another look?

ywelsch commented on core/src/main/java/org/elasticsearch/action/delete/DeleteResponse.java:
```diff
@@ -89,20 +59,44 @@ public boolean isFound() {
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(Fields.FOUND, isFound());
+        super.toXContent(builder, params);
```
switch these 2 lines so that id, version etc. appear first?

bleskes replied:
This how it used to be (found first), and I tend to agree - the found flag is the mode important one. Might be inconsistent but it's not my goal to change this in this PR (requires research on the other endpoints as well)

ywelsch replied:
ok, just noticed this in comparison to IndexResponse.

ywelsch commented on core/src/main/java/org/elasticsearch/rest/action/bulk/RestBulkAction.java:
```diff
@@ -97,49 +94,19 @@ public RestResponse buildResponse(BulkResponse response, XContentBuilder builder
                     if (itemResponse.isFailed()) {
+                        builder.field(Fields._INDEX, itemResponse.getIndex());
+                        builder.field(Fields._TYPE, itemResponse.getType());
+                        builder.field(Fields._ID, itemResponse.getId());
```
maybe we could go one step further and implement StatusToXContent on BulkItemResponse :-)

ywelsch commented on core/src/test/java/org/elasticsearch/update/UpdateIT.java:
```diff
@@ -64,14 +52,7 @@
-import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.Matchers.*;
 
 public class UpdateIT extends ESIntegTestCase {
```
can we remove this class from the commit? AFAICS only formatting changes.

ywelsch commented:
More lines removed than added -> I like this.
Left two suggestions, o.w. LGTM

martijnvg commented:
Good stuff. LGTM

bleskes commented:
@ywelsch pushed another round with your suggestions.

ywelsch commented:
LGTM
